easybuild/tools/modules.py

```diff
@@ -831,33 +834,54 @@ def available(self, mod_name=None):
-        cmd = [spider_cmd, '-o', 'moduleT', os.environ['MODULEPATH']]
-        self.log.debug("Running command '%s'..." % ' '.join(cmd))
+        """Update Lmod cache after new modules were added."""
+        cache_dir = build_option('update_lmod_caches')
 
-        proc = subprocess.Popen(cmd, stdout=PIPE, stderr=PIPE, env=os.environ)
-        (stdout, stderr) = proc.communicate()
+        if cache_dir is not None:
 
-        if stderr:
-            self.log.error("An error occured when running '%s': %s" % (' '.join(cmd), stderr))
+            for cache_type in ['moduleT', 'dbT', 'reverseMapT']:
 
-        if self.testing:
-            # don't actually update local cache when testing, just return the cache contents
-            return stdout
-        else:
-            try:
-                cache_filefn = os.path.join(os.path.expanduser('~'), '.lmod.d', '.cache', 'moduleT.lua')
-                self.log.debug("Updating Lmod spider cache %s with output from '%s'" % (cache_filefn, ' '.join(cmd)))
-                cache_dir = os.path.dirname(cache_filefn)
-                if not os.path.exists(cache_dir):
-                    mkdir(cache_dir, parents=True)
-                cache_file = open(cache_filefn, 'w')
-                cache_file.write(stdout)
-                cache_file.close()
-            except (IOError, OSError), err:
-                self.log.error("Failed to update Lmod spider cache %s: %s" % (cache_filefn, err))
+                # run spider to create cache contents
+                spider_cmd = os.path.join(os.path.dirname(self.cmd), 'spider')
+                cmd = [spider_cmd, '-o', cache_type, os.environ['MODULEPATH']]
+                self.log.debug("Running command '%s'..." % ' '.join(cmd))
+
+                # we can't use run_cmd, since we need to separate stdout from stderr
+                proc = subprocess.Popen(cmd, stdout=PIPE, stderr=PIPE, env=os.environ)
+                (cache_txt, stderr) = proc.communicate()
+                if stderr:
+                    self.log.error("An error occured when running '%s': %s" % (' '.join(cmd), stderr))
+
+                if self.testing:
+                    # don't actually update local cache when testing, just return the cache contents
+                    return cache_txt
+                else:
+                    cache_fp = os.path.join(cache_dir, '%s.lua' % cache_type)
+                    self.log.debug("Updating Lmod spider cache %s with output from '%s'" % (cache_fp, ' '.join(cmd)))
+
+                    # back existing cache by copying it to <cache>T.old.lua
+                    # Lmod also considers this filename for the cache in case <cache>T.lua isn't found
+                    if os.path.exists(cache_fp):
+                        old_cache_fp = '%s.old.lua' % os.path.splitext(cache_fp)[0]
+                        copy_file(cache_fp, old_cache_fp)
+                        self.compile_cache(old_cache_fp)
```

Comment:
ah, yes, but you're right, I should just move the old cache, rather than recompile is, just realised what you intended to point out